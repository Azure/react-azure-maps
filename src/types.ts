import { ComponentClass, CSSProperties, ReactElement, StatelessComponent } from 'react'
import atlas, {
  AnimationOptions,
  CameraBoundsOptions,
  CameraOptions,
  DataSourceOptions,
  HeatMapLayerOptions,
  HtmlMarker,
  HtmlMarkerEvents,
  HtmlMarkerOptions,
  ImageLayerOptions,
  LineLayerOptions,
  Map,
  MapDataEvent,
  MapErrorEvent,
  MapEvent,
  MapMouseEvent,
  MapMouseWheelEvent,
  MapTouchEvent,
  Options,
  PolygonExtrusionLayerOptions,
  PolygonLayerOptions,
  PopupEvents,
  PopupOptions,
  ServiceOptions,
  Shape,
  StyleOptions,
  SymbolLayerOptions,
  TargetedEvent,
  TileLayerOptions,
  TrafficOptions,
  UserInteractionOptions,
  Control,
  BubbleLayerOptions,
  LayerOptions,
  VectorTileSourceOptions
} from 'azure-maps-control'
import { drawing, DrawingManagerOptions, DrawingToolbarOptions } from 'azure-maps-drawing-tools'

export type IAzureMapOptions = ServiceOptions &
  StyleOptions &
  UserInteractionOptions &
  (CameraOptions | CameraBoundsOptions)

export type IAzureMapChildren =
  | ReactElement<IAzureMapHtmlMarker>
  | ReactElement<IAzureMapPopup>
  | ReactElement<IAzureMapDataSourceProps>

export type IAzureMap = {
  children?: Array<IAzureMapChildren> | IAzureMapChildren
  providedMapId?: string
  containerClassName?: string
  styles?: CSSProperties
  LoaderComponent?: ComponentClass<any> | StatelessComponent<any>
  options?: IAzureMapOptions
  imageSprites?: IAzureMapImageSprite[]
  controls?: IAzureMapControls[]
  customControls?: IAzureCustomControls[]
  events?: IAzureMapEvent | any
  cameraOptions?: AzureSetCameraOptions
  trafficOptions?: TrafficOptions
  userInteraction?: UserInteractionOptions
  styleOptions?: StyleOptions
  serviceOptions?: ServiceOptions
}
export type IAzureCustomControls = {
  control: Control
  controlOptions?: ControlOptions
}

export type IAzureMapControls = {
  controlName: string
  controlOptions?: Options
  options?: ControlOptions | undefined
}

export type IAzureMapImageSprite = {
  id: string
  templateName?: string
  color?: string
  secondaryColor?: string
  scale?: number
  icon?: string | HTMLImageElement | ImageData
}

export type IAzureMapContextState = {
  mapRef: Map | null
  isMapReady: boolean | false
  setMapRef(mapRef: Map): void
  removeMapRef(): void
  setMapReady(isMapReady: boolean): void
}

export type IAzureMapHtmlMarkerEvent = {
  eventName: keyof HtmlMarkerEvents
  callback: (e: TargetedEvent) => void
}

export type IAzureMapPopupEvent = {
  eventName: keyof PopupEvents
  callback: (e: TargetedEvent) => void
}

export type IAzureMapMouseEvents = {
  [T in keyof HtmlMarkerEvents]: (e: TargetedEvent) => void
}

export type IAzureMapHtmlMarker = {
  id?: string
  isPopupVisible?: boolean
  markerContent?: ReactElement
  options?: HtmlMarkerOptions
  events?: Array<IAzureMapHtmlMarkerEvent>
}

export type IAzureMapPopup = {
  isVisible?: boolean
  options?: PopupOptions
  events?: Array<IAzureMapPopupEvent>
  popupContent: ReactElement
}

export type IAzureMapDataSourceContextState = {
  dataSourceRef: atlas.source.DataSource | atlas.source.VectorTileSource | null
}

export type IAzureMapLayerContextState = {
  layerRef: atlas.layer.SymbolLayer | atlas.layer.ImageLayer | atlas.layer.TileLayer | null
}

export type IAzureDataSourceChildren =
  | IAzureMapFeature
  | ReactElement<IAzureMapFeature>
  | ReactElement<IAzureLayerStatefulProviderProps>

export type IAzureVectorTileSourceChildren = ReactElement<IAzureLayerStatefulProviderProps>

export type IAzureMapDataSourceEvent = {
  [property in IAzureMapDataSourceEventType]: (e: Shape[]) => void
}

export type IAzureMapVectorTileSourceEvent = {
  [property in IAzureMapSourceEventType]?: (e: atlas.source.VectorTileSource) => void
}

export type IAzureMapEvent = {
  [property in IAzureMapEventsType]: (
    e:
      | MapDataEvent
      | MapErrorEvent
      | MapTouchEvent
      | MapMouseEvent
      | string
      | MapMouseWheelEvent
      | MapEvent
      | atlas.layer.Layer
      | atlas.source.Source
  ) => void
}

export type IAzureDataSourceStatefulProviderProps = {
  id: string
  children?: | Array<IAzureDataSourceChildren | IAzureDataSourceChildren[] | null>
  | IAzureDataSourceChildren
  | null
  options?: DataSourceOptions
  events?: IAzureMapDataSourceEvent | any
  dataFromUrl?: string
  collection?:
    | atlas.data.FeatureCollection
    | atlas.data.Feature<atlas.data.Geometry, any>
    | atlas.data.Geometry
    | atlas.data.GeometryCollection
    | Shape
    | Array<atlas.data.Feature<atlas.data.Geometry, any> | atlas.data.Geometry | Shape>
  index?: number
}

export type IAzureVectorTileSourceStatefulProviderProps = {
  id: string,
  children?: | Array<IAzureVectorTileSourceChildren | IAzureVectorTileSourceChildren[] | null>
  | IAzureVectorTileSourceChildren
  | null
  options?: VectorTileSourceOptions,
  events?: IAzureMapVectorTileSourceEvent
  // NOTE: not sure yet why this is needed, haven't seen this used in AzureMapsDataSource, though IAzureGeoJSONDataSourceStatefulProviderProps has it
  index?: number
}

export type IAzureMapLayerEvent = {
  [property in IAzureMapLayerEventType]: (
    e: MapMouseEvent | MapTouchEvent | MapMouseWheelEvent
  ) => void
}

export type IAzureMapLifecycleEvent = {
  [property in IAzureMapLayerLifecycleEvents]: (e: atlas.layer.Layer) => void
}

export type IAzureLayerStatefulProviderProps = {
  id?: string
  options?:
    | (SymbolLayerOptions &
        HeatMapLayerOptions &
        ImageLayerOptions &
        LineLayerOptions &
        PolygonExtrusionLayerOptions &
        PolygonLayerOptions &
        TileLayerOptions &
        BubbleLayerOptions &
        LayerOptions)
    | Options
  type: IAzureMapLayerType
  events?: IAzureMapLayerEvent | any
  onCreateCustomLayer?: (dataSourceRef: DataSourceType, mapRef: MapType | null) => atlas.layer.Layer
  lifecycleEvents?: IAzureMapLifecycleEvent | any
}

export type IAzureDrawingLayerStatefulProviderProps = {
  // type of drawing layer to customize
  type: "lineLayer" | "polygonLayer" | "polygonOutlineLayer" | "pointLayer"
  options?:
    | (SymbolLayerOptions &
        LineLayerOptions &
        PolygonLayerOptions &
        LayerOptions)
    | Options
  events?: IAzureMapLayerEvent | any
  lifecycleEvents?: IAzureMapLifecycleEvent | any
}

export type IAzureMapLayerLifecycleEvents = 'layeradded' | 'layerremoved'

export type IAzureMapEventsType =
  | IAzureMapLayerEventType
  | IAzureMapLayerLifecycleEvents
  | IAzureMapDataSourceEventType
  | IAzureMapAddEventsType
  | IAzureMapSourceEventType
  // Adds a data event to the map.
  | 'data'
  | 'sourcedata'
  | 'styledata'
  // Adds an event to the map.
  | 'error'
  // Adds a style image missing event to the map.
  | 'styleimagemissing'

export type IAzureMapAddEventsType =
  | 'boxzoomstart'
  | 'boxzoomend'
  | 'dragstart'
  | 'drag'
  | 'dragend'
  | 'idle'
  | 'load'
  | 'movestart'
  | 'move'
  | 'moveend'
  | 'pitchstart'
  | 'pitch'
  | 'pitchend'
  | 'ready'
  | 'render'
  | 'resize'
  | 'rotatestart'
  | 'rotate'
  | 'rotateend'
  | 'tokenacquired'
  | 'zoomstart'
  | 'zoom'
  | 'zoomend'

export type IAzureMapDataSourceEventType = 'dataadded' | 'dataremoved'

export type IAzureMapSourceEventType = 'sourceadded' | 'sourceremoved'

export type IAzureMapLayerEventType =
  // Mouse events
  | 'mousedown'
  | 'mouseup'
  | 'mouseover'
  | 'mousemove'
  | 'click'
  | 'dblclick'
  | 'mouseout'
  | 'mouseenter'
  | 'mouseleave'
  | 'contextmenu'
  // Wheel events
  | 'wheel'
  // Touch events
  | 'touchstart'
  | 'touchend'
  | 'touchmove'
  | 'touchcancel'

export type IAzureMapLayerType =
  | 'SymbolLayer'
  | 'HeatLayer'
  | 'ImageLayer'
  | 'LineLayer'
  | 'PolygonExtrusionLayer'
  | 'PolygonLayer'
  | 'TileLayer'
  | 'BubbleLayer'
  | 'HtmlMarkerLayer'
  | 'custom'

export type IAzureMapFeatureType =
  | 'Point'
  | 'MultiPoint'
  | 'LineString'
  | 'MultiLineString'
  | 'Polygon'
  | 'MultiPolygon'

export type IAzureMapFeature = {
  id?: string
  type: IAzureMapFeatureType
  coordinate?: atlas.data.Position
  coordinates?: Array<atlas.data.Position>
  multipleCoordinates?: Array<Array<atlas.data.Position>>
  multipleDimensionCoordinates?: Array<Array<Array<atlas.data.Position>>>
  bbox?: atlas.data.BoundingBox
  variant?: IAzureMapFeatureVariant
  properties?: Options
  // Shape functions:
  setCoords?:
    | atlas.data.Position
    | atlas.data.Position[]
    | atlas.data.Position[][]
    | atlas.data.Position[][][]
  setProperties?: Options
}

// DRAWING MODULE:
export type DrawingManagerType = drawing.DrawingManager
export type IAzureDrawingManagerDrawingEventType = 
  | 'drawingchanged'
  | 'drawingchanging'
  | 'drawingcomplete'
  | 'drawingstarted'

export type IAzureDrawingManagerModeEventType = 'drawingmodechanged'
export type IAzureDrawingManagerEventType = IAzureDrawingManagerDrawingEventType | IAzureDrawingManagerModeEventType

export type IAzureDrawingManagerEvent = {
  [property in IAzureDrawingManagerDrawingEventType]?: (e: atlas.Shape) => void
} & {
  [property in IAzureDrawingManagerModeEventType]?: (a: drawing.DrawingMode) => void
}

export interface IAzureMapDrawingManagerProps {
  drawingManagerRef: drawing.DrawingManager | null
}

export interface IAzureDrawingManagerStatefulProviderProps {
  options: Omit<DrawingManagerOptions, 'toolbar'> & { 'toolbar'?: DrawingToolbarOptions }
  events?: IAzureDrawingManagerEvent
  children?: 
    | Array<IAzureDataSourceChildren | null> 
    | Array<ReactElement<IAzureDrawingLayerStatefulProviderProps> | null>
    | IAzureDataSourceChildren
    | ReactElement<IAzureDrawingLayerStatefulProviderProps>
    | null
}

export type IAzureMapLayerProps = IAzureMapLayerContextState
export type IAzureMapMouseEventRef = HtmlMarker // && other possible iterfaces
export type IAzureMapsContextProps = IAzureMapContextState
export type IAzureMapDataSourceProps = IAzureMapDataSourceContextState
export type DataSourceType = atlas.source.DataSource | atlas.source.VectorTileSource
export type LayerType = atlas.layer.SymbolLayer | atlas.layer.ImageLayer | atlas.layer.TileLayer
export type MapType = atlas.Map
export type GeometryType = atlas.data.Geometry
export type FeatureType = atlas.data.Feature<atlas.data.Geometry, any>
export type ShapeType = atlas.Shape
export type IAzureMapFeatureVariant = 'shape' | 'feature'

// Azure types
export type AzureDataLineString = atlas.data.LineString
export type AzureDataPosition = atlas.data.Position
export type ControlOptions = atlas.ControlOptions
export type AzureSetCameraOptions = ((CameraOptions | CameraBoundsOptions) & AnimationOptions) | any
