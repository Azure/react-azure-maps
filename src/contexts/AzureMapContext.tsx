import React, { createContext, ReactElement, useState ,useContext} from 'react'
import { Map } from 'azure-maps-control'
import { IAzureMap, IAzureMapsContextProps } from '../types'

export const AzureMapsContext = createContext<IAzureMapsContextProps>({
  mapRef: null,
  isMapReady: false,
  setMapRef: (mapRef: Map) => {},
  removeMapRef: () => {},
  setMapReady: () => {}
})
const { Provider, Consumer: AzureMapsConsumer } = AzureMapsContext

type IAzureMapsStatefulProviderProps = {
  children?: ReactElement<IAzureMap>
}

const AzureMapsStatefulProvider = ({ children }: IAzureMapsStatefulProviderProps) => {
  const [mapRef, setMapRef] = useState<Map | null>(null)
  const [isMapReady, setIsMapReady] = useState(false)

  return (
    <Provider
      value={{
        mapRef,
        setMapRef,
        isMapReady,
        setMapReady: setIsMapReady,
        removeMapRef: () => setMapRef(null)
      }}
    >
      {children}
    </Provider>
  )
}

const useAzureMaps = () => useContext<IAzureMapsContextProps>(AzureMapsContext)

export { AzureMapsConsumer, AzureMapsStatefulProvider as AzureMapsProvider, useAzureMaps }
