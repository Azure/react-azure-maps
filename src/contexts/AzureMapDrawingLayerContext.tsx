import React, { createContext } from 'react'
import { useAzureMapDrawingLayer } from '../hooks/useAzureMapDrawingLayer'
import { IAzureDrawingLayerStatefulProviderProps, IAzureMapLayerProps } from '../types'

const AzureMapDrawingLayerContext = createContext<IAzureMapLayerProps>({
  layerRef: null
})
const { Provider, Consumer: AzureMapDrawingLayerConsumer } = AzureMapDrawingLayerContext

const AzureMapDrawingLayerStatefulProvider = ({
  options,
  type,
  events,
  lifecycleEvents
}: IAzureDrawingLayerStatefulProviderProps) => {
  const { layerRef } = useAzureMapDrawingLayer({
    options,
    type,
    events,
    lifecycleEvents
  })

  return (
    <Provider
      value={{
        layerRef
      }}
    ></Provider>
  )
}

export {
  AzureMapDrawingLayerContext,
  AzureMapDrawingLayerConsumer,
  AzureMapDrawingLayerStatefulProvider as AzureMapDrawingLayerProvider
}
