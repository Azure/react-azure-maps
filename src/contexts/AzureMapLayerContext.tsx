import React, { createContext } from 'react'
import { IAzureLayerStatefulProviderProps, IAzureMapLayerProps } from '../types'
import { useAzureMapLayer } from '../hooks/useAzureMapLayer'

const AzureMapLayerContext = createContext<IAzureMapLayerProps>({
  layerRef: null
})
const { Provider, Consumer: AzureMapLayerConsumer } = AzureMapLayerContext

const AzureMapLayerStatefulProvider = ({
  id,
  options,
  type,
  events,
  lifecycleEvents,
  onCreateCustomLayer
}: IAzureLayerStatefulProviderProps) => {
  const { layerRef } = useAzureMapLayer({
    id,
    options,
    type,
    events,
    lifecycleEvents,
    onCreateCustomLayer
  })
  return (
    <Provider
      value={{
        layerRef
      }}
    ></Provider>
  )
}

export {
  AzureMapLayerContext,
  AzureMapLayerConsumer,
  AzureMapLayerStatefulProvider as AzureMapLayerProvider
}
