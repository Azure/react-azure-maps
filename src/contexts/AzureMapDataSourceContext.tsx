import React, { createContext, useContext, useEffect, useState } from 'react'
import {
  DataSourceType,
  IAzureDataSourceStatefulProviderProps,
  IAzureMapDataSourceProps,
  IAzureMapsContextProps,
  MapType
} from '../types'
import atlas from 'azure-maps-control'
import { AzureMapsContext } from './AzureMapContext'
import { useCheckRef } from '../hooks/useCheckRef'

const AzureMapDataSourceContext = createContext<IAzureMapDataSourceProps>({
  dataSourceRef: null
})
const { Provider, Consumer: AzureMapDataSourceConsumer } = AzureMapDataSourceContext

/**
 * @param id
 * @param children
 * @param options
 * @param events
 * @param dataFromUrl Async Data from URL
 * @param collection Use for contain Feature Collection !All Feature child will be ignored when collection value will change
 */
const AzureMapDataSourceStatefulProvider = ({
  id,
  children,
  options,
  events,
  dataFromUrl,
  collection
}: IAzureDataSourceStatefulProviderProps) => {
  const [dataSourceRef] = useState<atlas.source.DataSource>(new atlas.source.DataSource(id, options))
  const { mapRef } = useContext<IAzureMapsContextProps>(AzureMapsContext)
  useCheckRef<MapType, DataSourceType>(mapRef, dataSourceRef, (mref, dref) => {
    for (const eventType in events || {}) {
      mref.events.add(eventType as any, dref, events[eventType])
    }
    mref.sources.add(dref)
    if (dref instanceof atlas.source.DataSource) {
      if (dataFromUrl) {
        dref.importDataFromUrl(dataFromUrl)
      }
      if (collection) {
        dref.add(collection)
      }
    }
  })

  useEffect(() => {
    if (dataSourceRef && collection) {
      // Cleared old values prevent duplication
      dataSourceRef.clear()
      dataSourceRef.add(collection)
    }
  }, [collection])

  useEffect(() => {
    if (dataSourceRef && options) {
      dataSourceRef.setOptions(options)
    }
  }, [options])

  return (
    <Provider
      value={{
        dataSourceRef
      }}
    >
      {mapRef && children}
    </Provider>
  )
}

export {
  AzureMapDataSourceContext,
  AzureMapDataSourceConsumer,
  AzureMapDataSourceStatefulProvider as AzureMapDataSourceProvider,
  Provider as AzureMapDataSourceRawProvider
}
