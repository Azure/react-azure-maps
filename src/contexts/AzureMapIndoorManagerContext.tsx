import React, { createContext, useContext, useEffect, useState } from 'react'
import { indoor, control } from 'azure-maps-indoor'
import { IAzureMapIndoorManagerEventType, IAzureMapIndoorManagerStatefulProviderProps, IAzureMapsContextProps, MapType } from '../types'
import { AzureMapsContext } from './AzureMapContext'
import { useCheckRef } from '../hooks/useCheckRef'

const AzureMapIndoorManagerContext = createContext<{}>({ 
  indoorManagerRef: null 
})

const { Provider: IndoorManagerProvider, Consumer: AzureMapIndoorManagerConsumer } = AzureMapIndoorManagerContext

const AzureMapIndoorManagerStatefulProvider = ({
  options,
  children,
  dynamicStyling = false,
  facility = { facilityId: '', levelOrdinal: 0 },
  events = {}
}: IAzureMapIndoorManagerStatefulProviderProps) => {
  const { mapRef } = useContext<IAzureMapsContextProps>(AzureMapsContext)
  const [ indoorManagerRef, setIndoorManagerRef ] = useState<indoor.IndoorManager | null>(null)
  const [ levelControlRef, setLevelControlRef ] = useState<control.LevelControl | undefined>(undefined)

  useCheckRef<MapType, MapType>(mapRef, mapRef, mref => {
    mref.events.add('ready', () => {
      const indoorManager = new indoor.IndoorManager(mref)
      const levelControl = options.levelControl ? new control.LevelControl(options.levelControl) : undefined
      indoorManager.setOptions({
        ...options,
        levelControl
      })
      
      indoorManager.setDynamicStyling(dynamicStyling)
      indoorManager.setFacility(facility.facilityId, facility.levelOrdinal)

      setIndoorManagerRef(indoorManager)
      setLevelControlRef(levelControl)

      // TODO: gracefully handle the change of props events object
      for(const eventType in events){
        const handler = events[eventType as IAzureMapIndoorManagerEventType]
        if(!handler){ 
          continue 
        }

        mref.events.add(eventType as any, indoorManager, handler)
      }

      return () => {
        indoorManager.dispose()
        setIndoorManagerRef(null)
        setLevelControlRef(undefined)

        for(const eventType in events){
          const handler = events[eventType as IAzureMapIndoorManagerEventType]
          if(!handler){ 
            continue 
          }

          mref.events.remove(eventType, indoorManager, handler)
        }
      }
    })
  })

  useCheckRef(indoorManagerRef, options, (imref, options) => {
    imref.setOptions({
      ...options,
      levelControl: options.levelControl ? levelControlRef : undefined
    })
  })

  useCheckRef(levelControlRef, options.levelControl, () => {
    if(!options.levelControl){
      setLevelControlRef(undefined)
    } else {
      // NOTE: LevelControl does not yet support dynamically updating options yet 
      // levelControlRef.setOptions(options.levelControl)
      if(indoorManagerRef){
        const levelControl = new control.LevelControl(options.levelControl)
        setLevelControlRef(levelControl)
        indoorManagerRef.setOptions({
          ...options,
          levelControl
        })
      }
    }
  })

  useCheckRef(indoorManagerRef, dynamicStyling, (imref) => {
    imref.setDynamicStyling(dynamicStyling)
  })

  useCheckRef(indoorManagerRef, facility, (imref) => {
    imref.setFacility(facility.facilityId, facility.levelOrdinal)
  })

  return (
    <IndoorManagerProvider value={{indoorManagerRef}}>
      {mapRef && indoorManagerRef && children}
    </IndoorManagerProvider>
  )
}

export {
  AzureMapIndoorManagerContext,
  AzureMapIndoorManagerConsumer,
  AzureMapIndoorManagerStatefulProvider as AzureMapIndoorManagerProvider
}