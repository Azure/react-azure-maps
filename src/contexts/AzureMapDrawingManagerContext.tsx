import React, { createContext, useContext, useEffect, useState } from 'react'
import 'azure-maps-drawing-tools/dist/atlas-drawing.min.css'

import atlas from 'azure-maps-control'
import { drawing, control } from 'azure-maps-drawing-tools'
import {
  DrawingManagerType,
  IAzureDrawingManagerStatefulProviderProps,
  IAzureDrawingManagerEventType,
  IAzureDrawingManagerDrawingEventType,
  IAzureMapDrawingManagerProps,
  IAzureMapsContextProps,
  MapType
} from '../types'
import { AzureMapsContext } from './AzureMapContext'
import { useCheckRef } from '../hooks/useCheckRef'
import { Shape } from 'azure-maps-control'
import { AzureMapDataSourceContext } from './AzureMapDataSourceContext'

const AzureMapDrawingManagerContext = createContext<IAzureMapDrawingManagerProps>({
  drawingManagerRef: null
})

const { Provider: DrawingManagerProvider, Consumer: AzureMapDrawingManagerConsumer } = AzureMapDrawingManagerContext
const { Provider: DataSourceProvider } = AzureMapDataSourceContext

// FIXME: this is used until ts understanding azure-maps-control.EventManager extension in azure-maps-drawing-tools is resolved
// copies EventManager definitions from azure-maps-drawing-tools
interface _EventManager {
  add(eventType: "drawingmodechanged", target: drawing.DrawingManager, callback: (e: drawing.DrawingMode) => void): void;
  add(eventType: "drawingchanged" | "drawingchanging" | "drawingcomplete" | "drawingstarted", target: drawing.DrawingManager, callback: (e: Shape) => void): void;
  remove(eventType: string, target: drawing.DrawingManager, callback: (e?: any) => void): void
} 

const AzureMapDrawingManagerStatefulProvider = ({
  options,
  children,
  events = {}
}: IAzureDrawingManagerStatefulProviderProps) => {
  const { mapRef } = useContext<IAzureMapsContextProps>(AzureMapsContext)
  const [drawingManagerRef, setDrawingManagerRef] = useState<drawing.DrawingManager | null>(null)
  const [dataSourceRef, setDataSourceRef] = useState<atlas.source.DataSource | null>(null)
  const [toolbarOnceReadyRef, setToolbarOnceReadyRef] = useState<control.DrawingToolbar | undefined>(undefined)

  useCheckRef<MapType, MapType>(mapRef, mapRef, mref => {
    mref.events.add('ready', () => {
      // NOTE: if DrawingToolbar gets instantiated before map is 'ready', weird runtime errors follow. 
      // create toolbar here instead
      const drawingManager = new drawing.DrawingManager(mref)
      const toolbar = options.toolbar ? new control.DrawingToolbar(options.toolbar) : undefined
      drawingManager.setOptions({ 
        ...options, 
        toolbar
      })

      setDrawingManagerRef(drawingManager)
      setDataSourceRef(drawingManager.getSource())
      setToolbarOnceReadyRef(toolbar)
      
      // register drawing events
      for (const eventType in events) {
        const handler = events[eventType as IAzureDrawingManagerEventType]
        if(!handler){ 
          continue 
        }

        // NOTE: duplication to have the explicit types instead of any
        if(eventType == 'drawingmodechanged'){
          (mref.events as _EventManager).add(eventType, drawingManager, handler as (a: drawing.DrawingMode) => void)
        } else {
          (mref.events as _EventManager).add(eventType as IAzureDrawingManagerDrawingEventType, drawingManager, handler as (e: Shape) => void)
        }
      }

      return () => {
        setDrawingManagerRef(null)
        setDataSourceRef(null)
        setToolbarOnceReadyRef(undefined)

        for (const eventType in events) {
          const handler = events[eventType as IAzureDrawingManagerEventType];
          if(handler){
            (mref.events as _EventManager).remove(eventType, drawingManager, handler)
          }
        }
      }
    })
  })

  useEffect(() => {
    if(drawingManagerRef && options){
      drawingManagerRef.setOptions({ ...options, toolbar: toolbarOnceReadyRef })
    }
    if(toolbarOnceReadyRef && options && options.toolbar){
      toolbarOnceReadyRef.setOptions(options.toolbar)
    }
  }, [drawingManagerRef, options, toolbarOnceReadyRef])

  return (
    <DrawingManagerProvider
      value={{
        drawingManagerRef
      }}
    >
      <DataSourceProvider value={{ dataSourceRef }}>
        {mapRef && drawingManagerRef && children}
      </DataSourceProvider>
    </DrawingManagerProvider>
  )
}

export {
  AzureMapDrawingManagerContext,
  AzureMapDrawingManagerConsumer,
  AzureMapDrawingManagerStatefulProvider as AzureMapDrawingManagerProvider
}
