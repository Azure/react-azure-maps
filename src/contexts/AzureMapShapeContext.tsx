import React, { createContext, useContext, useEffect, useState } from 'react'
import {
  DataSourceType,
  IAzureLayerStatefulProviderProps,
  IAzureMapDataSourceProps,
  IAzureMapLayerProps,
  IAzureMapsContextProps,
  IAzureMapShapeProps,
  IAzureShapeStatefulProviderProps,
  MapType
} from '../types'
import { useAzureMapLayer } from '../hooks/useAzureMapLayer'
import atlas, { Shape } from 'azure-maps-control'
import { AzureMapsContext } from './AzureMapContext'
import { useCheckRef } from '../hooks/useCheckRef'
import { AzureMapDataSourceContext } from './AzureMapDataSourceContext'

const AzureMapShapeContext = createContext<IAzureMapShapeProps>({
  shapeRef: null
})

const { Provider, Consumer: AzureMapShapeConsumer } = AzureMapShapeContext

const AzureMapShapeStatefulProvider = ({
  id,
  properties,
  children
}: IAzureShapeStatefulProviderProps) => {
  const { mapRef } = useContext<IAzureMapsContextProps>(AzureMapsContext)
  const { dataSourceRef } = useContext<IAzureMapDataSourceProps>(AzureMapDataSourceContext)
  const [shapeRef, setShapeRef] = useState<atlas.Shape | any>(null)

  const createShape = (
    data: atlas.data.Feature<atlas.data.Geometry, any> | atlas.data.Geometry | any
  ) => {
    if (data) {
      const shape = new atlas.Shape(data, id, properties)
      setShapeRef(shape)
    }
  }

  useCheckRef<DataSourceType, Shape>(dataSourceRef, shapeRef, (dsref, sref) => {
    dsref.add(sref)
  })

  return (
    <Provider
      value={{
        shapeRef,
        setShapeRef,
        createShape
      }}
    >
      {mapRef && dataSourceRef && children}
    </Provider>
  )
}

export {
  AzureMapShapeContext,
  AzureMapShapeConsumer,
  AzureMapShapeStatefulProvider as AzureMapShapeProvider
}

// new atlas.Shape(data: atlas.data.Geometry, id, options)
