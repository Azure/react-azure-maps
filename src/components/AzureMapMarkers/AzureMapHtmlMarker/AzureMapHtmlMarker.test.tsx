import { render } from '@testing-library/react'
import { Map, Popup } from 'azure-maps-control'
import React from 'react'
import { AzureMapsContext } from '../../../contexts/AzureMapContext'
import AzureMaphtmlMarker from './AzureMapHtmlMarker'
import { IAzureMapHtmlMarker } from '../../../types'
import atlas from 'azure-maps-control'

const mapContextProps = {
  mapRef: null,
  isMapReady: false,
  setMapReady: jest.fn(),
  removeMapRef: jest.fn(),
  setMapRef: jest.fn()
}
const mapRef = new Map('fake', {})

const wrapWithAzureMapContext = (props: IAzureMapHtmlMarker) => {
  return (
    <AzureMapsContext.Provider
      value={{
        ...mapContextProps,
        mapRef
      }}
    >
      <AzureMaphtmlMarker {...{ ...props }} />
    </AzureMapsContext.Provider>
  )
}

describe('AzureMaphtmlMarker tests', () => {
  it('should create html marker without error and add it to map ref', () => {
    mapRef.markers.add = jest.fn()
    const { container, unmount } = render(wrapWithAzureMapContext({}))
    expect(mapRef.markers.add).toHaveBeenCalled()
    expect(container).toMatchSnapshot()
    unmount()
    expect(mapRef.markers.remove).toHaveBeenCalled()
  })

  it('should remove marker from map ref', () => {
    mapRef.markers.remove = jest.fn()
    const { unmount } = render(wrapWithAzureMapContext({}))
    unmount()
    expect(mapRef.markers.remove).toHaveBeenCalled()
  })

  it('should add events for marker to map', () => {
    mapRef.events.add = jest.fn()
    render(
      wrapWithAzureMapContext({
        events: [
          {
            eventName: 'click',
            callback: jest.fn()
          }
        ]
      })
    )
    expect(mapRef.events.add).toHaveBeenCalledWith(
      'click',
      expect.any(Object),
      expect.any(Function)
    )
  })
  describe('options and content change', () => {
    const markerRef = new atlas.HtmlMarker()
    // We need somehow to override current mock constructor ts currently do not allow that
    // @ts-ignore
    atlas.HtmlMarker = jest.fn(() => markerRef)

    it('should call setOptions on markerRef', () => {
      markerRef.setOptions = jest.fn()
      render(wrapWithAzureMapContext({ options: { option: 'option' } }))
      expect(markerRef.setOptions).toHaveBeenCalledWith({ option: 'option' })
    })

    it('should call setOptions on markerRef when marketContent prop is passed', () => {
      markerRef.setOptions = jest.fn()
      render(wrapWithAzureMapContext({ markerContent: <div /> }))
      expect(markerRef.setOptions).toHaveBeenCalledWith({ htmlContent: '<div></div>' })
    })

    it('should close marker popup', () => {
      render(wrapWithAzureMapContext({ isPopupVisible: true, markerContent: <div /> }))
      // Currently we only check if getOptions get called @TODO
      expect(markerRef.getOptions).toHaveBeenCalled()
    })

    it('should open marker popup', () => {
      render(wrapWithAzureMapContext({ isPopupVisible: false, markerContent: <div /> }))
      // Currently we only check if getOptions get called @TODO
      expect(markerRef.getOptions).toHaveBeenCalled()
    })
  })
})
