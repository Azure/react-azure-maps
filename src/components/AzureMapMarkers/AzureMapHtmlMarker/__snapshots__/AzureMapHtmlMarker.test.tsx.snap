// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AzureMaphtmlMarker tests should create html marker without error and add it to map ref 1`] = `<div />`;
